crates/sui-core/src/authority_active/gossip/node_sync.rs

```diff
@@ -0,0 +1,435 @@
+                }
+            });
+        }
+
+        let _ = rx
+            .recv()
+            .await
+            .map_err(|e| SuiError::GenericAuthorityError {
+                error: format!("{:?}", e),
+            })??;
+
+        self.node_sync_store
+            .get_cert_and_effects(&digest)?
+            .ok_or_else(|| SuiError::GenericAuthorityError {
+                error: format!(
+                    "cert/effects for {:?} should have been in the node_sync_store",
+                    digest
+                ),
+            })
+    }
+
+    async fn download_impl(
+        peer: AuthorityName,
+        aggregator: Arc<AuthorityAggregator<A>>,
+        digests: &ExecutionDigests,
+        node_sync_store: Arc<NodeSyncStore>,
+    ) -> SuiResult {
+        let digest = digests.transaction;
+
+        // TODO: Add a function to AuthorityAggregator to try multiple validators - even
+        // though we are fetching the cert/effects from the same validator that sent us the tx
+        // digest, and even though we know the cert is final, any given validator may be byzantine
+        // and refuse to give us the cert and effects.
+        let client = aggregator.clone_client(&peer);
+        let resp = client
+            .handle_transaction_and_effects_info_request(digests)
+            .await
+            .expect("TODO: need to use authority aggregator to download cert");
+
+        let cert = resp.certified_transaction.ok_or_else(|| {
+            info!(?digest, ?peer, "validator did not return cert");
+            SuiError::GenericAuthorityError {
+                error: format!("validator did not return cert for {:?}", digest),
+            }
+        })?;
+
+        let effects = resp.signed_effects.ok_or_else(|| {
+            info!(?digest, ?peer, "validator did not return effects");
+            SuiError::ByzantineAuthoritySuspicion { authority: peer }
+        })?;
+
+        node_sync_store.store_cert_and_effects(&digest, &(cert, effects))?;
+
+        Ok(())
+    }
+}
+
+#[derive(Clone)]
+pub struct NodeSyncDigestHandler {
+    _sync_join_handle: Arc<JoinHandle<()>>,
+    sender: mpsc::Sender<DigestsMessage>,
+}
+
+impl NodeSyncDigestHandler {
+    pub fn new<A>(
+        state: Arc<AuthorityState>,
+        aggregator: Arc<AuthorityAggregator<A>>,
+        node_sync_store: Arc<NodeSyncStore>,
+    ) -> Self
+    where
+        A: AuthorityAPI + Send + Sync + 'static + Clone,
+    {
+        let (sender, receiver) = mpsc::channel(NODE_SYNC_QUEUE_LEN);
+
+        let committee = state.committee.load().deref().clone();
+        let sync_state = Arc::new(NodeSyncState {
+            committee,
+            effects_stake: Mutex::new(EffectsStakeMap::new()),
+            state,
+            aggregator,
+            node_sync_store,
+            pending_downloads: Waiter::new(),
+            pending_txes: Waiter::new(),
+        });
+
+        let _sync_join_handle = Arc::new(NodeSyncState::start(sync_state, receiver));
```

Comment:
Yeah the actual NodeSyncState probably doesn't need to be wrapped in an arc since you're not holding references to it outside of the actual task and you're communicating with it via a channel.

Reply:
Removing this extra layer of wrapping will let you remove the Mutexes inside of the task and should simplify and speed up some things

Reply:
It won't, see above: We want concurrent downloads, so the Mutexes are necessary.